### PR TITLE
Add SkyWalking go SDK as a trace plugin

### PR DESCRIPTION
**Pull request by arugal**
### Issues associated with this PR
#1027 

Your PR should present related issues you want to solve.

### Solutions
You should show your solutions about the issues in your PR, including the overall solutions, details and the changes. At this time, Chinese is allowed to describe these.

### UT result
Unit Test is needed if the code is changed, your unit test should cover boundary cases, corner cases, and some exceptional cases. And you need to show the UT result.

### Benchmark
If your code involves the processing of every request, you should give the Benchmark Result.

### Code Style
+ Make sure `Goimports` has run
+ Show `Golint` result

### TODO list
1. [x] [go2sky](https://github.com/SkyAPM/go2sky/pull/49) releases an exclusive version of mosn
1. [x] Add the componentId and UI icon to [skywalking](https://github.com/apache/skywalking) repo
1. [x] test with skywalking

### CODE REVIEW

**Comment by arugal** [comment]
### Screenshot

#### Service
image

#### Endpoint
image

#### Instance
image

#### Topo
image

#### Trace
image
image
image

**Comment by arugal** [comment]
@wu-sheng @taoyuanyuan @rootsongjc please review :)

File: pkg/trace/skywalking/http/tracer.go [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+	entry, nCtx, err := tracer.CreateEntrySpan(ctx, requestURI, func() (sw6 string, err error) {
+		for _, h := range sw6Header {
+			sw6, ok = header.Get(h)
+			if ok {
+				// delete the sw6 header, otherwise the upstream service will receive two sw6 header
+				header.Del(h)
+				return sw6, err
+			}
+		}
+		return
```
[the remark this answers is not in the document]
Comment (arugal) [reply pkg/trace/skywalking/http/tracer.go]: Hi @wangfakang  Thank you for your review. The downstream service does not necessarily use the skywalking probe, so not `ok` is normal :)

File: pkg/trace/skywalking/http/tracer.go [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+		if requestInfo.ResponseCode() > 400 {
+			exit.Error(time.Now(), skywalking.ErrorLog)
+		}
+		exit.Tag(go2sky.TagStatusCode, responseCode)
+		exit.End()
+	}
+
+	// entry span (downstream)
+	entry := h.carrier.entrySpan
+	if requestInfo.ResponseCode() > 400 {
```
[the remark this answers is not in the document]
Comment (arugal) [reply pkg/trace/skywalking/http/tracer.go]: done

**Comment by arugal** [comment]
@taoyuanyuan @rootsongjc Did I miss any documentation?

**Comment by wu-sheng** [comment]
@arugal Thank you for adding this quickly.

**Comment by rootsongjc** [comment]
> @taoyuanyuan @rootsongjc Did I miss any documentation?

Do you have any usage instructions or manual?

File: pkg/trace/skywalking/http/tracer.go [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+			}
+		}
+		return
+	})
+	if err != nil {
+		log.DefaultLogger.Errorf("[SkyWalking] [tracer] [http1] create entry span error, err: %v", err)
+		return skywalking.NoopSpan
+	}
+	entry.Tag(go2sky.TagHTTPMethod, string(header.Method()))
+	entry.Tag(go2sky.TagURL, string(header.Host())+requestURI)
```
Comment (taoyuanyuan) [inline pkg/trace/skywalking/http/tracer.go:82]: Host maybe is nil that HTTP1.0， it‘s ok？

File: pkg/trace/skywalking/span.go [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package skywalking
+
+import (
+	"mosn.io/api"
```
Comment (taoyuanyuan) [inline pkg/trace/skywalking/span.go:21]: goimports

File: pkg/trace/skywalking/http/tracer.go [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package http
+
+import (
```
Comment (taoyuanyuan) [inline pkg/trace/skywalking/http/tracer.go:20]: goimports

File: pkg/trace/skywalking/tracer.go [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package skywalking
+
+import (
+	"encoding/json"
```
Comment (taoyuanyuan) [inline pkg/trace/skywalking/tracer.go:21]: goimports

File: pkg/trace/skywalking/span.go (hunk 4 above)
Comment (arugal) [reply pkg/trace/skywalking/span.go:23]: done

File: pkg/trace/skywalking/http/tracer.go (hunk 3 above)
Comment (arugal) [reply pkg/trace/skywalking/http/tracer.go]: I've changed the code.

File: pkg/trace/skywalking/http/tracer.go [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+	}
+	entry.Tag(go2sky.TagHTTPMethod, string(header.Method()))
+	// https://github.com/mosn/mosn/pull/1061#discussion_r401571554
+	host := header.Header()
+	var url string
+	if host != nil {
+		url = string(host) + requestURI
+	} else {
+		url = requestURI
+	}
```
Comment (taoyuanyuan) [inline pkg/trace/skywalking/http/tracer.go:90]: I'm sorry, I don't express clearly. the Host maybe is ni slice, the previous code that string(host) is right.
I mean that whether trace is affected by no host？

Comment (arugal) [reply pkg/trace/skywalking/http/tracer.go:90]: >  I mean that whether trace is affected by no host？

Should not affect, here is just tag. 
image

Comment (taoyuanyuan) [reply pkg/trace/skywalking/http/tracer.go:90]: ok, please rollback to the previous code, It’s more clearly.
thank you.

Comment (arugal) [reply pkg/trace/skywalking/http/tracer.go]: done

**Comment by arugal** [comment]
> > @taoyuanyuan @rootsongjc Did I miss any documentation?
> 
> Do you have any usage instructions or manual?

Hi @rootsongjc , I have improved the sample document, please review :)

File: examples/cn_readme/trace/skywalking/http/README.md [hunk 8]
```diff
@@ -0,0 +1,70 @@
+## 使用 SkyWalking 作为 trace 实现
```
Comment (rootsongjc) [inline examples/cn_readme/trace/skywalking/http/README.md:1]: 这里应该使用一级标题。

File: examples/cn_readme/trace/skywalking/http/README.md [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+
+## 简介
+
++ 该样例工程演示了如何配置使得[SkyWalking](http://skywalking.apache.org/)作为MOSN的trace实现
```
Comment (rootsongjc) [inline examples/cn_readme/trace/skywalking/http/README.md:5]: 句尾加句号。

File: examples/cn_readme/trace/skywalking/http/README.md [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+
+## 准备
+
++ 安装docker & docker-compose
+
+1. [安装docker](https://docs.docker.com/install/) <br>
+1. [安装docker-compose](https://docs.docker.com/compose/install/)
+
++ 需要一个编译好的MOSN程序
+```
```
Comment (rootsongjc) [inline examples/cn_readme/trace/skywalking/http/README.md:15]: 所有代码片段指定编程语言。

File: examples/cn_readme/trace/skywalking/http/README.md [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+
+### 启动MOSN
+
++ 使用config.json 运行MOSN
+
+```bash
+./main start -c config.json
+```
+
+### 启动一个Http Client
```
Comment (rootsongjc) [inline examples/cn_readme/trace/skywalking/http/README.md:64]: 为啥这么多标题，普通的描述没必要用标题了。

File: examples/cn_readme/trace/skywalking/http/README.md (hunk 9 above)
Comment (arugal) [reply examples/cn_readme/trace/skywalking/http/README.md]: done

File: examples/cn_readme/trace/skywalking/http/README.md (hunk 10 above)
Comment (arugal) [reply examples/cn_readme/trace/skywalking/http/README.md]: done

File: examples/cn_readme/trace/skywalking/http/README.md (hunk 11 above)
Comment (arugal) [reply examples/cn_readme/trace/skywalking/http/README.md]: done

File: examples/cn_readme/trace/skywalking/http/README.md (hunk 8 above)
Comment (arugal) [reply examples/cn_readme/trace/skywalking/http/README.md]: done

**Comment by rootsongjc** [comment]
This [PR](https://github.com/mosn/mosn/pull/1061) has been merged. It will be released in the next version of MOSN.

**Pull request merged**